domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -161,101 +190,291 @@ class TopicListController @Inject constructor(
+
+      val isTopicConsideredCompleted = checkIfAtLeastOneStoryIsCompleted(topicProgress, topic)
+
       topicProgress.storyProgressMap.values.forEach { storyProgress ->
         val storyId = storyProgress.storyId
         val story = topicController.retrieveStory(topic.topicId, storyId)
 
-        val completedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
-        val lastCompletedChapterProgress: ChapterProgress? =
+        val completedChapterProgressList = getCompletedChapterProgressList(storyProgress)
+        val mostRecentCompletedChapterProgress: ChapterProgress? =
           completedChapterProgressList.firstOrNull()
 
-        val startedChapterProgressList =
-          storyProgress.chapterProgressMap.values
-            .filter { chapterProgress ->
-              chapterProgress.chapterPlayState ==
-                ChapterPlayState.STARTED_NOT_COMPLETED
-            }
-            .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
-
-        val recentlyPlayerChapterProgress: ChapterProgress? =
+        val startedChapterProgressList = getStartedChapterProgressList(storyProgress)
+        val latestStartedChapterProgress: ChapterProgress? =
           startedChapterProgressList.firstOrNull()
-        if (recentlyPlayerChapterProgress != null) {
-          val recentlyPlayerChapterSummary: ChapterSummary? =
-            story.chapterList.find { chapterSummary ->
-              recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
-            }
-          if (recentlyPlayerChapterSummary != null) {
-            val numberOfDaysPassed =
-              (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
-            val promotedStory = createPromotedStory(
+
+        when {
+          latestStartedChapterProgress != null -> {
+            createOngoingStoryListBasedOnRecentlyPlayed(
               storyId,
+              story,
+              latestStartedChapterProgress,
+              startedChapterProgressList,
               topic,
-              completedChapterProgressList.size,
-              story.chapterCount,
-              recentlyPlayerChapterSummary.name,
-              recentlyPlayerChapterSummary.explorationId
-            )
-            if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
-              ongoingStoryListBuilder.addRecentStory(promotedStory)
-            } else {
-              ongoingStoryListBuilder.addOlderStory(promotedStory)
+              isTopicConsideredCompleted
+            ).let {
+              Pair(it.first, it.second)
+              if (completionTimeFilter(it.second))
+                it.first?.let { it1 -> playedPromotedStoryList.add(it1) }
```

Comment:
Here & below: suggest naming ``it1`` to provide context  (e.g. ``daysPassed``).

Reply:
Removed this statement.